nes-input-formatters/private/STBufferState.hpp

```diff
@@ -0,0 +1,195 @@
+
+namespace NES::InputFormatters
+{
+
+using STBufferIdx = NESStrongType<uint32_t, struct STBufferIdx_, 0, 1>;
+using ABAItNo = NESStrongType<uint32_t, struct ABAItNo_, 0, 1>;
+
+class STBufferEntry;
+
+/// (Implementation detail of the STBufferEntry)
+/// The AtomicState consists a 64-bit-wide atomic bitmap that allows the STBuffer to thread-safely determine whether a specific
+/// STBufferEntry may overwrite a prior entry or whether it connects to a neighbour-entry and whether the connecting entries form an ST
+/// Central to the AtomicState are the first 32 bits, which represent the 'ABA iteration number' (abaItNo)
+/// The abaItNo protects against the ABA problem, but more importantly, it allows the STBuffer to determine which iteration an STBufferEntry
+/// belongs to (see header documentation of STBuffer)
+/// Furthermore, the bitmap denotes whether an entry (with a specific abaItNo) has a delimiter, whether it was claimed as the first buffer
+/// in a spanning tuple, and whether the leading/trailing buffer of the STBufferEntry were moved out already
+/// If both the 'usedLeadingBufferBit' and the 'usedTrailingBufferBit' are set, then the buffer has no more uses and it is safe to replace
+/// it with the buffer that has a sequence number that corresponds to the same STBufferEntry and that has the next higher abaItNo
+/// (Planned) Some of the (currently) 28 unused bits will be used to model uncertainty. If a delimiter is larger than a byte a thread may
+///           be uncertain whether the first/last few bytes of its buffer form a valid tuple delimiter. Additionally, it may be impossible
+///           for a thread to determine if its buffer starts in an escape sequence or not. We can model these additional pieces of
+///           information as states using the remaning bits and (potentially) perform state transitions as (atomic) bitwise operations.
+class AtomicState
+{
+    friend STBufferEntry;
+
+    /// 33:   000000000000000000000000000000100000000000000000000000000000000
+    static constexpr uint64_t hasTupleDelimiterBit = 4294967296ULL;
+    /// 34:   000000000000000000000000000001000000000000000000000000000000000
+    static constexpr uint64_t claimedSpanningTupleBit = 8589934592ULL;
+    /// 35:   000000000000000000000000000010000000000000000000000000000000000
+    static constexpr uint64_t usedLeadingBufferBit = 17179869184ULL;
+    /// 36:   000000000000000000000000000100000000000000000000000000000000000
+    static constexpr uint64_t usedTrailingBufferBit = 34359738368ULL;
+    ///       000000000000000000000000000110000000000000000000000000000000000
+    static constexpr uint64_t usedLeadingAndTrailingBufferBits = 51539607552ULL;
+
+    /// The STBuffer initializes all STBufferEntries, except for the very first entry, with the 'defaultState'
+    /// Tag: 0, HasTupleDelimiter: True, ClaimedSpanningTuple: True, UsedLeading: True, UsedTrailing: True
+    static constexpr uint64_t defaultState = 60129542144ULL;
```

Comment:
Can this not be expressed with the constexpr of above?
something like: 
```suggestion
    static constexpr uint64_t defaultState = hasTupleDelimiter | claimedSpanningTupleBit ... ;
```

Reply:
did exactly that and then replaced the rest with bitshiftts

Reply:
you should see them now